### PR TITLE
cephadm: add host-maintenance subcommand

### PR DESCRIPTION
**Pull request by pcuzner**
Initial patch to provide an enter and exit subcommand
to cephadm. Note these are simply 'primitives' that will
be used by mgr/cephadm

Signed-off-by: Paul Cuzner <pcuzner@redhat.com>

### CODE REVIEW

**Comment by pcuzner** [comment]
maintenance-mode-demo

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -6861,6 +6929,18 @@ def _get_parser():
         help='daemon identifer for the exporter')
     parser_exporter.set_defaults(func=command_exporter)
 
+    parser_maintenance = subparsers.add_parser(
+        'host-maintenance', help='Manage the maintenance state of a host')
+    parser_maintenance.add_argument(
+        '--fsid',
+        help='cluster FSID')
+    parser_maintenance.add_argument(
```
Comment (jmolmo) [inline src/cephadm/cephadm:6937]: we will need to add an option "--check" parameter to the "enter" command. Maybe we will need use separate arguments for each "choice"

**Review by jmolmo — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
greenish: https://pulpito.ceph.com/swagner-2020-12-23_18:12:01-rados:cephadm-wip-swagner-testing-2020-12-22-0110-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
`maintenance` is implemented by disabling  `ceph.target`. The idea to have this comment really makes sense, but to me, this looks dangerous. 

We really want to call `ok-to-stop` before stopping Ceph services to be on the same side and this is already implemented! 

https://github.com/ceph/ceph/blob/dd9f4a9345b2d2c7590c23a3591b0f0c537a0756/src/pybind/mgr/cephadm/services/cephadmservice.py#L208

Thus, you really want to call this, *before* enabling the maintenance mode. 

https://github.com/ceph/ceph/blob/dd9f4a9345b2d2c7590c23a3591b0f0c537a0756/src/pybind/mgr/cephadm/module.py#L1373

But at that time, you're **already** in the MGR module, and thus don't need to invent new hooks. 

Why not trivially add a new command to mgr/cephadm that calls `host_ok_to_stop` and then calls

https://github.com/ceph/ceph/blob/dd9f4a9345b2d2c7590c23a3591b0f0c537a0756/src/pybind/mgr/cephadm/module.py#L1601

for each daemon on this host with `action="stop"`?

**Comment by pcuzner** [comment]
@sebastian-philipp I know what you mean, but maintenance is generally a hardware cycle that has to support multiple reboots (BIOS, hardware changes etc). If you simply stop the daemons, and the machine restarts the ceph daemons will start again - that's why I opted for a disable at the systemd level to ensure no matter what the admin is doing we don't end up with services starting when then shouldn't. The way I've implemented the feature means that until the admin removes maintenance, you can't get any erroneous service starts.

**Comment by sebastian-philipp** [comment]
Understood, but if you're unlucky, stopping the services this way is might render the cluster in a read-only mode. 

That's why ceph-salt is making sure, the host is ok-to-stop, before running those maintenance steps:

https://github.com/ceph/ceph-salt/blob/7d9e307af3da011d65516e68be1d25d97b28ecdb/ceph-salt-formula/salt/_states/ceph_orch.py#L180

I agree that you probably want to disable the ceph.target to avoid starting the services, but you probably don't want to use the ceph.target to stop the services.

**Comment by pcuzner** [comment]
But host_ok_to_stop is being invoked, if the host has daemons on it - so I'm not sure what the objection is here?
https://github.com/pcuzner/ceph/blob/2beb87881179d868a1c42dc3956f676261f45051/src/pybind/mgr/cephadm/module.py#L1465

**Comment by sebastian-philipp** [comment]
relates to https://tracker.ceph.com/issues/48779

**Pull request merged**